pkg/cmd/server/kubernetes/master_config_test.go

```diff
@@ -1,14 +1,130 @@
 package kubernetes
 
 import (
+	"net"
 	"reflect"
 	"testing"
+	"time"
 
+	apiserveroptions "k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	cmapp "k8s.io/kubernetes/cmd/kube-controller-manager/app/options"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
 	"k8s.io/kubernetes/pkg/genericapiserver"
+	kubeletclient "k8s.io/kubernetes/pkg/kubelet/client"
+	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/util"
 
 	configapi "github.com/openshift/origin/pkg/cmd/server/api"
 )
 
+func TestAPIServerDefaults(t *testing.T) {
+	defaults := apiserveroptions.NewAPIServer()
+
+	// This is a snapshot of the default config
+	// If the default changes (new fields are added, or default values change), we want to know
+	// Once we've reacted to the changes appropriately in BuildKubernetesMasterConfig(), update this expected default to match the new upstream defaults
+	expectedDefaults := &apiserveroptions.APIServer{
+		ServerRunOptions: &genericapiserver.ServerRunOptions{
+			BindAddress:          net.ParseIP("0.0.0.0"),
+			CertDirectory:        "/var/run/kubernetes",
+			InsecureBindAddress:  net.ParseIP("127.0.0.1"),
+			InsecurePort:         8080,
+			LongRunningRequestRE: "(/|^)((watch|proxy)(/|$)|(logs?|portforward|exec|attach)/?$)",
+			SecurePort:           6443,
+		},
+		APIGroupPrefix:          "/apis",
+		APIPrefix:               "/api",
+		AdmissionControl:        "AlwaysAdmit",
+		AuthorizationMode:       "AlwaysAllow",
+		DeleteCollectionWorkers: 1,
+		EnableLogsSupport:       true,
+		EtcdConfig: etcdstorage.EtcdConfig{
+			Prefix: "/registry",
+		},
+		EventTTL:               1 * time.Hour,
+		MasterCount:            1,
+		MasterServiceNamespace: "default",
+		RuntimeConfig:          util.ConfigurationMap{},
+		StorageVersions:        registered.AllPreferredGroupVersions(),
+		DefaultStorageVersions: registered.AllPreferredGroupVersions(),
+		KubeletConfig: kubeletclient.KubeletClientConfig{
+			Port:        10250,
+			EnableHttps: true,
+			HTTPTimeout: time.Duration(5) * time.Second,
+		},
+	}
+
+	if !reflect.DeepEqual(defaults, expectedDefaults) {
```

Comment:
No test will be perfect here, but this test would have caught the things that I've cared about in the past (new exposed ports for cadvisor, changed network proxy default modes, etc). JSON serialization wouldn't help if things had omitempty JSON tags, I also want compile errors. Would rather improve in a follow up if we think it's necessary.